### PR TITLE
Feature/using GitHub pull request template

### PR DESCRIPTION
**Pull request by 4383**
Using pull request template file for PR message when is detected in project sources.
Users are invited to fill the template in defined system editor ($EDITOR)

### CODE REVIEW

File: git_pull_request/__init__.py [hunk 1]
```diff
@@ -149,6 +151,13 @@ def git_get_title_and_message(begin, end):
     else:
         title = "Pull request for " + end
         message = "\n".join(titles)
+
+    pr_template = find_pull_request_template()
```
Comment (jd) [inline git_pull_request/__init__.py]: This should be done L 147 directly, no need to do the lines above.

File: git_pull_request/__init__.py [hunk 2]
```diff
@@ -245,21 +254,56 @@ def download_pull_request(g, repo, target_remote, pull_number):
...
+        content = body.read().strip()
+    os.unlink(bodyfilename)
+    return content
+
+
+def edit_title_and_message(title, message):
+    pr_template = find_pull_request_template()
+    if pr_template:
+        content = get_pr_template_message(pr_template)
+        return title, content
```
Comment (jd) [inline git_pull_request/__init__.py]: This is wrong, this is called when editing a pull request and in this case you don't want to override the content of the existing PR with the template.

File: git_pull_request/tests/__init__.py [hunk 3]
```diff
@@ -129,7 +169,7 @@ def test_only_title(self):
             gpr.parse_pr_message("foobar\n\nsomething\nawesome\n"))
 
 
-class TestMessageEditor(unittest.TestCase):
+class TestMessageEditor(fixtures.TestWithFixtures):
```
Comment (jd) [inline git_pull_request/tests/__init__.py]: Does not seem related?

File: git_pull_request/__init__.py [hunk 4]
```diff
@@ -134,6 +136,18 @@ def git_get_log_titles(begin, end):
     return list(split_and_remove_empty_lines(log))
 
 
+def generate_title(titles, end):
+    if len(titles) == 1:
+        return titles[0]
+    return "Pull request for " + end
+
+
+def generate_message(titles, end):
```
Comment (jd) [inline git_pull_request/__init__.py]: both these are only used once so not sure it's worth creating them TBH

File: git_pull_request/__init__.py [hunk 5]
```diff
@@ -142,13 +156,15 @@ def git_get_title_and_message(begin, end):
...
     """
     titles = git_get_log_titles(begin, end)
-
-    if len(titles) == 1:
-        title = titles[0]
-        message = git_get_commit_body(end)
+    title = generate_title(titles, end)
+    pr_template = find_pull_request_template()
+    if pr_template:
+        LOG.warning(
```
Comment (jd) [inline git_pull_request/__init__.py]: there's no need for a warning here

File: git_pull_request/__init__.py [hunk 6]
```diff
@@ -245,21 +261,51 @@ def download_pull_request(g, repo, target_remote, pull_number):
...
+    return content
+
+
+def edit_title_and_message(title, message):
     fd, bodyfilename = tempfile.mkstemp()
     with open(bodyfilename, "w") as body:
         body.write(title + "\n\n")
         body.write(message + "\n")
-    status = os.system(editor + " " + bodyfilename)
+    status = os.system(get_editor() + " " + bodyfilename)
```
Comment (jd) [inline git_pull_request/__init__.py]: rather than doing a `get_editor()` function, do a `edit_file` function that you can use here and in `get_pr_template_message`.

**Comment by jd** [comment]
I like the fact that you contribute to git-pull-request without using it it seems ;)

**Comment by 4383** [comment]
@jd tester is doubting ;)
squash and refactor done!

**Comment by 4383** [comment]
applying your last review done!

File: git_pull_request/__init__.py [hunk 7]
```diff
@@ -245,21 +252,48 @@ def download_pull_request(g, repo, target_remote, pull_number):
...
+def edit_title_and_message(title, message):
     fd, bodyfilename = tempfile.mkstemp()
     with open(bodyfilename, "w") as body:
         body.write(title + "\n\n")
         body.write(message + "\n")
-    status = os.system(editor + " " + bodyfilename)
-    if status != 0:
-        raise RuntimeError("Editor exited with status code %d" % status)
+    edit_file(bodyfilename)
+
```
Comment (jd) [inline git_pull_request/__init__.py]: You can also factorize the read/unlink part IMHO.

just rename to `edit_file_get_content_and_remove` or something like that if you wish ;)

Comment (4383) [reply git_pull_request/__init__.py]: done!
